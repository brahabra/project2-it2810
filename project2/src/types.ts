export type Commit = { 
    id: string,
    short_id: string,
    created_at: string,
    author_email: string, 
    author_name: string,
    authored_date: string, 
    committed_date: string,
    committer_email: string, 
    committer_name: string,
    message: string,
    title: string, 
    web_url: string,
};
