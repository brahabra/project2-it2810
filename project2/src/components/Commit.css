.commit {
    width: 200px;
    background-color: beige;
    border-style: solid;
    padding: 20px;
    margin: auto;
}

.commit .title {
    font-size: 20px;
}
