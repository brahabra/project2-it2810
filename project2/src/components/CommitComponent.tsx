import { Container, Paper } from "@mui/material";
import { Commit } from "../types"
import { style } from "../styles/Styles";

export const CommitComponent = (props: {key:string, commit:Commit}) => {
  return (
    <Container>
      <Paper sx={style.commitComponentPaper}>
        <p><strong>TITLE: </strong> <a href={props.commit.web_url} > {props.commit.title}</a>
        <br/>
        <strong>AUTHOR: </strong>  {props.commit.committer_name}
        <br/>
        <strong>COMMITTER EMAIL: </strong>{props.commit.committer_email}
        <br/>
        <strong>DATE CREATED: </strong>{props.commit.created_at.substring(0, 10)}</p>
      </Paper>
    </Container>
  );
};
