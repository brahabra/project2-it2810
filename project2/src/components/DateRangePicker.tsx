import * as React from "react";
import { Dayjs } from "dayjs";
import TextField from "@mui/material/TextField";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { DatePicker } from "@mui/x-date-pickers/DatePicker";

interface Props{
    startValue: Dayjs | null,
    setStartValue: (value: Dayjs | null) => void,
    endValue: Dayjs | null,
    setEndValue: (value: Dayjs | null) => void
  }


export default function DateRangePicker(
    props: Props
  ) {
 
  return (
    <LocalizationProvider dateAdapter={AdapterDayjs}>
      <DatePicker
        label="Start date"
        value={props.startValue}
        onChange={(newStartValue) => {
          if (newStartValue != null) {
            props.setStartValue(newStartValue);
          }
        }}
        renderInput={(params) => <TextField {...params} />}
      />
      <DatePicker
        label="End date"
        value={props.endValue}
        onChange={(newEndValue) => {
          if (newEndValue != null) {
            props.setEndValue(newEndValue);
          }
        }}
        renderInput={(params) => <TextField {...params} />}
      />
    </LocalizationProvider>
  );
}
