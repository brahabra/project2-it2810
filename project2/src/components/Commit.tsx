import React, { useState } from "react";
import { Typography } from "@mui/material";
import "../styles/Commit.css";


type Commit = { 
  id: string,
  short_id: string,
  created_at: string,
  author_email: string, 
  author_name: string,
  authored_date: string, 
  committed_date: string,
  committer_email: string, 
  committer_name: string,
  message: string,
  title: string,
  web_url: string,
}

export const Commit = (props: {key:string, commit:Commit}) => {
  return (
    <div className="commit">
      <Typography>
        <div className="title"><a href={props.commit.web_url}>{props.commit.title}</a></div>
        <div className="comitter">@{props.commit.committer_name} </div>
        <div className="created">{props.commit.created_at.substring(0, 10)}</div>
      </Typography>
    </div>
  );
};
