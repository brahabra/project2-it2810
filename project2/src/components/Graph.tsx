import React, { useState } from "react";
import { Container, Paper, Typography, IconButton } from "@mui/material";
import "../styles/Commit.css";
import { Commit } from "../types"
import { PieChart } from 'react-minimal-pie-chart';

function getColor(memberNr: number) {
  const startRange:number = 65280;
  const endRange:number = 16777216;
  const intervall:number = 5000;
  let color:string = "#" + ((startRange+intervall*memberNr)%endRange).toString(16);
  while (color.length < 7) {
    color = color + "0"
  }
  return color;
}

type Member = {
  name: string;
  commits: number;
  color: string;
};

function calcMemberDistrobution(commits: Commit[]) {
  let members: Map<string, Member> = new Map();
  let member: string;
  let memberNr:number = 0;
  for (let i = 0; i < commits.length; i++) {
    member = commits[i].committer_name;
    if (typeof members.get(member) === "undefined") {
      members.set(member, {name: member, commits: 0, color: getColor(memberNr)})
      memberNr++; 
    } 
    members.get(member)!.commits++;  
  }
  let data: Member[] = [];
  members.forEach((value) => {
     data.push(value);
  })
  return data;
}

interface Props {
    commits: Commit[],
}

export const Graph = (props: Props) => {
    const membersDistrobution = calcMemberDistrobution(props.commits);
    const chartData: {title: string, value:number, color: string}[] = [];
    let member: Member;

    for (let i = 0; i < membersDistrobution.length; i++) {
      member = membersDistrobution[i];
      chartData.push( { title: member.name, value: member.commits, color: member.color } );
    }
  return (
    <Container>
      <Paper sx={{
            padding: '20px',
            borderRadius: '20px',
            width: 'auto',
            margin: '20px',
            textAlign: 'left'
          }}
      >
        <>
        <PieChart
            data={chartData}
        />
          {membersDistrobution.map(((member: Member)=>(
            <Typography key={member.name} sx={{
                                          display: 'flex',
                                          flexDirection: 'row',
            }}>
              <div>
                <svg width="15" height="15">
                  <rect width="10" height="10" rx="2" fill={member.color}/>
                </svg>
              </div>
               <div>{member.name}</div>
               </Typography>
          )))}
        </>
      </Paper>
    </Container>
  );
};