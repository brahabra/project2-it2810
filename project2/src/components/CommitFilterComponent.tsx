import { Container, Box } from "@mui/system";
import { Dayjs } from "dayjs";
import { useState, useEffect } from "react";
import { style } from "../styles/Styles";
import { Commit } from "../types";
import { SessionStorageClass } from "../WebStorageClass";
import DateRangePicker from "./DateRangePicker";
import NameRowComponent from "./NameRowComponent";

//Properties interface for filter component.
interface Props {
  commits: Commit[];
  setFilterList: (value: Commit[]) => void;
  filterList: Commit[];
}

//Component for selecting name on commits page. This filters out all commits not made by selected 
//person.
export const FilterComponent = (props: Props) => {
  const [filterName, setName] = useState("");
  const [startValue, setStartValue] = useState<Dayjs | null>(null);
  const [endValue, setEndValue] = useState<Dayjs | null>(null);

  //Loading previos selected member (if the member has commits) from session storage, 
  //when this page loads.
  useEffect(() => {
    function hasCommits(name:string | null) {
      if (name === null) {
        return false;
      }
      for (let c of props.commits) {
        if (c.committer_name == name) {
            return true
        } 
      }
      return false
    }
    const storage = new SessionStorageClass();
    const name = storage.getPropValue("selectedName");
    if (hasCommits(name)) {
      setName(name!);
    } else {
      setName("");
    }
  }, []);

  //The function that filters commits. Runs on updates to commits list, filterName variable
  //(the name that is selected), start date and end date for calender.
  useEffect(() => {
    function filterCommitList() {
      const commitsCopy = [...props.commits];
      let filteredCommits: Commit[] = commitsCopy;
      if (filterName && filterName !== "default") {
        filteredCommits = filteredCommits.filter(
          (commit) => commit.author_name === filterName
        );
      }

      if (startValue) {
        filteredCommits = filteredCommits.filter(
          (commit) =>
            startValue.isBefore(commit.created_at) ||
            startValue.isSame(commit.created_at, "day")
        );
      }

      if (endValue) {
        filteredCommits = filteredCommits.filter(
          (commit) =>
            endValue.isAfter(commit.created_at) ||
            endValue.isSame(commit.created_at, "day")
        );
      }
      props.setFilterList(filteredCommits);
    }

    filterCommitList();
  }, [props.commits, filterName, startValue, endValue]);

  return (
    <Container>
      <Box sx={style.commitSelectDates}>
        <DateRangePicker
          startValue={startValue}
          setStartValue={setStartValue}
          endValue={endValue}
          setEndValue={setEndValue}
        />
      </Box>
      <Box sx={style.commitSelectName}>
      <NameRowComponent
        filterName={filterName}
        commits={props.commits}
        setName={setName}
      />
      </Box>
    </Container>
  );
};
