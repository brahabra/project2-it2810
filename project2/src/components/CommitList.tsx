import { Box, Container } from "@mui/material";
import { useContext, useEffect, useState } from "react";
import { getCommits } from "../api/fetch";
import { ProjectContext } from "../ProjectContext";
import { Commit } from "../types";
import { CommitComponent } from "./CommitComponent";
import { FilterComponent } from "./CommitFilterComponent";
import { Graph } from "./Graph";
import { style } from "../styles/Styles";
import "../styles/CommitsPage.css";

/*interface Props {
    selectedBranch: string
}*/

export const CommitList = () => {
    const ctx = useContext(ProjectContext);
    const [commits, setCommits] = useState<Commit[]>([]);
    const [filterList, setFilterList] = useState<Commit[]>([]);
    const [isLoading, setLoading] = useState(false);


    useEffect(() => {
      getCommits(ctx.projectID, ctx.token /*, props.selectedBranch*/).then(
        (res: Commit[]) => {
          setCommits(res);
          setLoading(true);
        }
      );
    }, [ /*props.selectedBranch,*/ ctx.projectID, ctx.token]);

    return (
        <div className="commitContainer">
          {isLoading ? <Graph commits={commits} /> : null} 
          <Box sx={style.commitListBox}>
          {isLoading ? <FilterComponent commits={commits} filterList={filterList} setFilterList={setFilterList}/> : null }
            {isLoading ? 
            <Box sx={style.commitListContainer}>
                {filterList.map((commit) => (
                    <CommitComponent key={commit.id} commit={commit} />
                ))}
            </Box> : <p>Loading Data ...</p> }
            </Box>
        </div>

    );

    
}