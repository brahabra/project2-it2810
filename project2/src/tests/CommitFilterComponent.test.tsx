import * as React from 'react'
import { FilterComponent } from '../components/CommitFilterComponent';
import { Commit } from '../types';

const ReactTestRenderer = require('react-test-renderer');

const commit: Commit = {
  id: "123",
  short_id: "123",
  created_at: "2022-09-30",
  author_email: "test@example.com", 
  author_name: "test test",
  authored_date: "2022-09-30", 
  committed_date: "2022-09-30",
  committer_email: "test@example.com", 
  committer_name: "test test",
  message: "test test",
  title: "test commit", 
  web_url: "example.com",
} 

it('renders correctly', () => {
  const setFilterList = jest.fn();
  const tree = ReactTestRenderer.create(<FilterComponent commits={[commit]} filterList={[]} setFilterList={setFilterList} />).toJSON()
  expect(tree).toMatchSnapshot()
})