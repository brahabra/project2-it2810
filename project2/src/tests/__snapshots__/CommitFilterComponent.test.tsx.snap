// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="dateRange"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r0:"
        id=":r0:-label"
      >
        Start date
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="mm/dd/yyyy"
          required={false}
          type="tel"
          value=""
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              Start date
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r1:"
        id=":r1:-label"
      >
        End date
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="mm/dd/yyyy"
          required={false}
          type="tel"
          value=""
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              End date
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="MuiBox-root css-ece9u5"
  >
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
      >
        Name
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1ftyaf0"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
